### ISSUE TITLE
NeuralynxIO questions

### ISSUE THREAD

**Issue by emelon8** [NeuralEnsemble/python-neo#868]
Hi,

I have some questions about NeuralynxIO:

1. I noticed that blocks of the raw signal with less than 512 valid samples are discarded. Is there a reason for this? From my conversations with Neuralynx technical support, it seems that it would be better to keep all valid samples, even if they come from a chunk with fewer than 512 valid samples.

For questions 2 and 3, assume
```
reader = neo.io.NeuralynxIO(recording_directory)
block = reader.read_block(signal_group_mode='split-all')
```

2. Does `block.segments[0].analogsignals[0].t_stop` correspond to the time the last sample in that segment was collected? I am guessing it does not, since `block.segments[0].analogsignals[0].times[-1]` is one timestep shorter.

3. If I do `reader.segment_t_start(0,0)`, I get 0.0 s, whereas if I do `block.segments[0].analogsignals[0].t_start`, I get 0.001969 s (I am sampling at 500 Hz). Does this mean that the first analog signal sample was collected 0.001969 s after the start of the block? The reason I ask is because I want to make sure I know how to align `block.segments[0].events[0].times` with the analog signals. For example, would an event with a timestamp of 0.001969 s have occurred at the exact same time as the first analog signal sample (as opposed to 0.001969 s after the first sample)?

4. The analog signal data imported by Neuralynx's MATLAB import function looks like it has the opposite sign as the same data imported into Python with Neo (i.e., multiplied by -1). Is there a reason for this?

5. Is there a way, using neo, to write events back to .nev files? For instance, if I want to correct typos in the event labels or add/delete events?

Thanks for any help you can provide, and thanks for developing a great tool!

-Eric

**Comment by samuelgarcia** [comment]
I did the refactoring of the code but never use the neuralynxio that I don't have perfect knownledge on this.

1. I don't remember that we skip small blocks. @JuliaSprenger ?

2. AnalogSignal is responsible of the t_stop behavior. I think the actual behavior make sens. Have a look to this:

 ```
    @property
    def duration(self):
        '''
        Signal duration

        (:attr:`size` * :attr:`sampling_period`)
        '''
        return self.shape[0] / self.sampling_rate

    @property
    def t_stop(self):
        '''
        Time when signal ends.

        (:attr:`t_start` + :attr:`duration`)
        '''
        return self.t_start + self.duration
```


3. A segment can have several object analogsignal and events. I guess that some objects (event) are starting at 0s.
Whereas the very first sample of analogsignal is not 0. 
In neo the object model reveal the fact that not all objects start at the same time.
About neuralynxio fine implementation about timestamp I don't known if this is correct. I guess yes.

4. Neuralynx have internaly a flag that tell that the signal is inverted. I guess matlab implementation do not deal with it.
```
                    if info['input_inverted']:
                        gain *= -1
```

5. Neuralynxio do not have write ablity. If you want to clean and save back your objects you should use NixIO for instance.

**Comment by JuliaSprenger** [comment]
Hi @emelon8!
To answer question 1: As far as I know the file specifications for Neuralynx ncs files state that data blocks always contain 512 samples, see [this format specification](https://neuralynx.com/software/NeuralynxDataFileFormats.pdf). This is why we consider blocks with less samples as incomplete and potentially corrupted and ignore these to ensure only valid data is loaded. This implies that whenever there is a corrupted block, the continuous data will be handled as two separate recordings that are represented by different segments within the neo structure. Ignoring the missing samples instead and concatenating all blocks would lead to incorrect time stamps of samples after the incomplete block.

For question 3: The relative timing of data within a segment is being taken care of by the NeuralynxIO. So if your event has the same time stamp as a sample of your Analogsignal, this means there were also reported as occurring simultaneously by Neuralynx. However, as each segment has it's own time frame,  time stamps between segments are not necessary aligned / comparable.

I hope this helps. Let us know if this raises further questions!

**Comment by emelon8** [comment]
Hi Julia and Samuel,

Thank you both for your detailed answers! Definitely makes me want to continue using Neo.

**Regarding Question 1**: When I was discussing the issue of data blocks with less than 512 valid samples with Neuralynx Support earlier this year, they sent me some sample MATLAB code one of their engineers wrote for how to deal with this problem, which I've edited for clarity and posted below.

`Timestamps` is an array of the timestamps for the first sample of each block. It is of size N x 1.
`ts` is the time step.
`Samples` is an array of all the sampled data in the recordings. It is of size 512 x N.
`NumberOfValidSamples` is an array of the number of valid samples in each block (512 or fewer). It is of size N x 1. Thus, the length of the entire recording is `sum(NumberOfValidSamples)`, unless you add samples to fill in gaps between segments.

```
% Prepare vector structures to store sample data and timestamps to return
sample_data = zeros(sum(NumberOfValidSamples),1);
time = zeros(sum(NumberOfValidSamples),1);

% Creates vector of sample data and corresponding timestamps from CSC recording. 
start_index = 1; % points to start of new record
for i = 1:length(NumberOfValidSamples)
    
    % fills sample_data vector with number of valid samples of record
    sample_data(start_index: start_index+NumberOfValidSamples(i)-1) = Samples(1:NumberOfValidSamples(i),i);
    
    % fills time over number of valid samples.
    time(start_index: start_index+NumberOfValidSamples(i)-1) = Timestamps(i): ts: Timestamps(i)+(NumberOfValidSamples(i)-1)*ts;
    
    % increments start_index to next record
    start_index = start_index + NumberOfValidSamples(i);
end
```

I hope that's clear! As you can see, they keep all valid samples, even from blocks with fewer than 512 valid samples. In the case of NeuralynxIO, valid samples from blocks with fewer than 512 valid samples would be at the end of their respective segment. Then, a new segment would start with the next block as is currently implemented. The difference between the two cases looks like this: ![incomplete_record_example](https://user-images.githubusercontent.com/8397842/94203580-aa589f80-fe8d-11ea-92b2-087e600065d1.png)
The signals are the same, except for where there are fewer than 512 valid samples. Can we implement a way to keep all valid samples in NeuralynxIO?

**Comment by JuliaSprenger** [comment]
Hi @emelon8 ,
yes, in principle we can extent the NeuralynxIO to also include the samples of incomplete data blocks if you are sure that the Neuralynx Support confirmed that the existing samples of an incomplete block are always valid. Did they confirm that your case is not a special case and there might be other cases of incomplete blocks where the samples shouldn't be treated as valid?

**Comment by sin-mike** [comment]
> 1. I noticed that blocks of the raw signal with less than 512 valid samples are discarded. Is there a reason for this? From my conversations with Neuralynx technical support, it seems that it would be better to keep all valid samples, even if they come from a chunk with fewer than 512 valid samples.

Hi Eric. I made this patch. The reason for the patch is keeping the consistency of the lazy (==memmap reading in the NLX case) interfaces in the reader without overheading in RAM. It is still possible to reconstruct valid patches, but I'm kind of away from the coding at the moment myself. You can also follow the discussion in PR #693. 

The quick and RAMy solution would be to make a mask for each block concerning its valid counts. The non-trivial solution would be to make an interface function for the validation of faulty blocks at the moment of reading. I'm not sure I could handle this up to the new year's.

**Comment by emelon8** [comment]
Hi @JuliaSprenger ,
The code that the Neuralynx engineer provided was not written for my particular case. It was just a MATLAB script for CSC data that contains incomplete records that was shared with me when I asked how to handle them in my analyses. I reached out to them again and they confirmed that we "can safely keep valid samples of a CSC record even though they are less than 512 samples."

Hi @sin-mike ,
Being somewhat of a beginner with Neo (and, to a lesser extent, Python), I'm not sure I understand what you are proposing. Is there a way I can access the blocks with fewer than 512 valid samples within the existing NeuralynxIO object and recreate the signal (with _all_ valid samples) myself? I only see blocks with 512 samples in `recording._sigs_memmap`.

**Comment by samuelgarcia** [comment]
@emelon8 
In the actual state the code is pretty fast : it can select relevant block because they have all the same size 512/
It it were not the case it would be more slow with some tricj like cumsum.
See https://github.com/NeuralEnsemble/python-neo/blob/master/neo/rawio/neuralynxrawio.py#L300-L323

If you have in mind a way to keep shorter block and keep the fast get_analogsignal_chunk, please propose a PR.
Maybe it is possible because the shorter is always the last one in a Segment.

**Comment by emelon8** [comment]
@samuelgarcia 
I dug into it more and found that the valid samples can be added at the end of each segment with some minor edits to neuralynxrawio.py. I created a pull request to make the necessary changes.

**Comment by emelon8** [comment]
It seems that my edits are failing the test in line 310 of test_neuralynxio.py. I wonder if the t_stops (8427.830803, 8487.768029, 8515.816549) that it compares nio._sigs_t_stop to in lines 309-310 need to be updated, because including the valid samples in the last record of a segment results in different t_stops than before.

**Comment by JuliaSprenger** [comment]
This seems to be indirectly closed via #881
